### PR TITLE
[Dynamo] Support torch.{cuda/cpu}.amp.autocast

### PR DESCRIPTION
**Pull request by yanboliang**
For Meta internal use cases.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/torch.py [hunk 1]
```diff
@@ -324,6 +324,13 @@ def call_function(
             )
         elif self.value is torch.amp.autocast_mode.autocast:
             return AutocastModeVariable.create(target_values=args, kwargs=kwargs)
+        elif self.value in [torch.cuda.amp.autocast, torch.cpu.amp.autocast]:
+            assert "device_type" not in kwargs
```
Comment (jansel) [inline torch/_dynamo/variables/torch.py:328]: Should this be `unimplemented()` to graph break rather than assert?

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/variables/torch.py (hunk 1 above)
Comment (yanboliang) [reply torch/_dynamo/variables/torch.py:328]: eager mode throwing error for this case:
```
Traceback (most recent call last):
  File "/scratch/ybliang/work/repos/pytorch/debug/debug2.py", line 29, in <module>
    print(fn(x))
  File "/scratch/ybliang/work/repos/pytorch/debug/debug2.py", line 20, in f2
    with torch.cuda.amp.autocast(device_type="cuda", dtype=torch.bfloat16):
TypeError: __init__() got an unexpected keyword argument 'device_type'
```
So Dynamo should throw ```AssertionError```?

**Comment by yanboliang** [comment]
Currently blocked by https://github.com/pytorch/pytorch/issues/95837

**Comment by weiwangmeta** [comment]
Real failure is https://github.com/pytorch/pytorch/actions/runs/4360278495/jobs/7623296244#step:11:1070 
not the 
image

**Comment by huydhn** [comment]
Here is the error snippet:

```
+ python benchmarks/dynamo/check_memory_compression_ratio.py --actual /var/lib/jenkins/workspace/test/test-reports/inductor_training_smoketest_hf_Albert.csv --expected benchmarks/dynamo/expected_ci_perf_inductor_torchbench.csv

            hf_Albert                         :
                actual_memory_compression=1.19,
                expected_memory_compression=1.26,
                FAIL
            

Error: 1 models below expected memory compression ratio:
    hf_Albert
If this drop is expected, you can update `benchmarks/dynamo/expected_ci_perf_inductor_torchbench.csv`.
```

**Comment by davidberard98** [comment]
This caused a number of dynamo + eager backend accuracy failures on the timm suite (tested regnety_002, but possibly others) cc @Chillee to investigate

**Comment by yanboliang** [comment]
@davidberard98 thanks for your investigation. This PR actually fixed a serious bug in dynamo benchmark: for all AMP benchmarks, actually we fallback to eager mode before this PR. It may have some metric changes since we truly support ```torch.cuda.autocast``` in dynamo and benchmarks, and definitely we should identify where is the failure from firstly.

**Pull request closed without merging**